File: src/environment/Environment.ts
/** Represents an execution environment. */
export interface Environment {
    log(text: string): void;
    warn(text: string): void;
    error(text: string): void;
    resolvePath(path: string): string;
    readFile(filePath: string): Promise<string>;
    writeFile(filePath: string, text: string): Promise<void>;
    glob(patterns: string[]): Promise<string[]>;
}

File: packages/rust-dprint/src/environment/environment.rs
use std::path::PathBuf;

pub trait Environment : std::marker::Sync {
    fn read_file(&self, file_path: &PathBuf) -> Result<String, String>;
    fn write_file(&self, file_path: &PathBuf, file_text: &str) -> Result<(), String>;
    fn log(&self, text: &str);
    fn log_error(&self, text: &str);
}
